src/lib.rs

```diff
@@ -38,72 +34,136 @@ impl rowan::Language for NixLanguage {
-
-        let let_in = ast.root().inner().and_then(LetIn::cast).unwrap();
-        let set = let_in.body().and_then(AttrSet::cast).unwrap();
-        let inherit = set.inherits().nth(1).unwrap();
-
-        let from = inherit.from().unwrap().inner().and_then(Ident::cast).unwrap();
-        assert_eq!(from.to_inner_token().text(), "set");
-        let mut children = inherit.idents();
-        assert_eq!(children.next().unwrap().to_inner_token().text(), "z");
-        assert_eq!(children.next().unwrap().to_inner_token().text(), "a");
-        assert!(children.next().is_none());
-    }
-    #[test]
-    fn math() {
-        let ast = parse(include_str!("../test_data/general/math.nix"));
-        let root = ast.root().inner().and_then(BinOp::cast).unwrap();
-        let operation = root.lhs().and_then(BinOp::cast).unwrap();
+}
 
-        assert_eq!(root.operator().unwrap(), BinOpKind::Add);
-        assert_eq!(operation.operator().unwrap(), BinOpKind::Add);
+/// The result of a parse
+#[derive(Clone)]
+pub struct Parse<T> {
+    green: GreenNode,
+    errors: Vec<ParseError>,
+    _ty: PhantomData<fn() -> T>,
+}
+
+impl<T> Parse<T> {
+    pub fn syntax(&self) -> SyntaxNode {
+        SyntaxNode::new_root(self.green.clone())
+    }
+}
 
-        let lhs = operation.lhs().and_then(Value::cast).unwrap();
-        assert_eq!(lhs.to_value(), Ok(NixValue::Integer(1)));
+impl<T: AstNode> Parse<T> {
+    pub fn tree(&self) -> T {
+        T::cast(self.syntax()).unwrap()
+    }
 
-        let rhs = operation.rhs().and_then(BinOp::cast).unwrap();
-        assert_eq!(rhs.operator().unwrap(), BinOpKind::Mul);
+    /// Return all errors in the tree, if any
+    pub fn errors(&self) -> &[ParseError] {
+        &*self.errors
     }
-    #[test]
-    fn t_macro() {
-        assert_eq!(T![@], SyntaxKind::TOKEN_AT);
-        assert!(matches!(SyntaxKind::TOKEN_PAREN_OPEN, T!["("]));
+
+    /// Either return the first error in the tree, or if there are none return self
+    pub fn ok(self) -> Result<T, ParseError> {
+        if let Some(err) = self.errors().first() {
+            return Err(err.clone());
+        }
+        Ok(self.tree())
     }
 }
+
+/// Matches a `SyntaxNode` against an `ast` type.
+///
+/// # Example:
+///
+/// ```ignore
+/// match_ast! {
+///     match node {
+///         ast::CallExpr(it) => { ... },
+///         ast::MethodCallExpr(it) => { ... },
+///         ast::MacroCall(it) => { ... },
+///         _ => None,
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! match_ast {
+    (match $node:ident { $($tt:tt)* }) => { match_ast!(match ($node) { $($tt)* }) };
+
+    (match ($node:expr) {
+        $( ast::$ast:ident($it:pat) => $res:expr, )*
+        _ => $catch_all:expr $(,)?
+    }) => {{
+        $( if let Some($it) = ast::$ast::cast($node.clone()) { $res } else )*
+        { $catch_all }
+    }};
+}
+
+#[cfg(test)]
+mod tests {
+    // #[test]
+    // fn interpolation() {
```

Comment:
This test checks that interpolation and dedention work correctly when interleaved. As it needs to check the resulting AST (because errors in the past happened in the AST-forming process), it should be ported.